libs/@local/harpc/server/src/router.rs

```diff
@@ -0,0 +1,244 @@
+use alloc::sync::Arc;
+use core::{
+    future::{self, Ready},
+    task::{Context, Poll},
+};
+
+use frunk::{HCons, HNil};
+use futures::future::Either;
+use harpc_net::codec::ErrorEncoder;
+use harpc_service::delegate::ServiceDelegate;
+use harpc_tower::{
+    body::{Body, BodyExt},
+    request::Request,
+    response::{BoxedResponse, Parts, Response},
+};
+use harpc_types::{service::ServiceId, version::Version};
+use tower::{Layer, Service, ServiceBuilder, ServiceExt, layer::util::Identity, util::Oneshot};
+
+use crate::{delegate::ServiceDelegateHandler, error::NotFound, session::SessionStorage};
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct Handler<S> {
+    service: ServiceId,
+    version: Version,
+
+    inner: S,
+}
+
+pub trait Route<B, C> {
+    type ResponseBodyError;
+    type Future: Future<Output = Result<BoxedResponse<Self::ResponseBodyError>, !>> + Send;
+
+    fn call(&self, request: Request<B>, codec: C) -> Self::Future
+    where
+        B: Body<Control = !, Error: Send + Sync> + Send + Sync,
+        C: ErrorEncoder + Send + Sync;
+}
+
+// The clone requirement might seem odd here, but is the same as in axum's router implementation.
+// see: https://docs.rs/axum/latest/src/axum/routing/route.rs.html#45
```

Comment:
It's not helpful to compare with other crates and what they are using. Why the `Clone` is needed?

Reply:
see: [1e966ef](https://github.com/hashintel/hash/pull/5320/commits/1e966efcdacabb7210df7f00d2f6e94f25168973)

Reply:
Much better, thank you!